// @osaas/client-core/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.listInstances = exports.getInstance = exports.removeInstance = exports.createInstance = exports.Context = exports.createFetch = exports.Log = void 0;
var log_1 = require("./log");
Object.defineProperty(exports, "Log", { enumerable: true, get: function () { return log_1.Log; } });
var fetch_1 = require("./fetch");
Object.defineProperty(exports, "createFetch", { enumerable: true, get: function () { return fetch_1.createFetch; } });
var context_1 = require("./context");
Object.defineProperty(exports, "Context", { enumerable: true, get: function () { return context_1.Context; } });
var core_1 = require("./core");
Object.defineProperty(exports, "createInstance", { enumerable: true, get: function () { return core_1.createInstance; } });
Object.defineProperty(exports, "removeInstance", { enumerable: true, get: function () { return core_1.removeInstance; } });
Object.defineProperty(exports, "getInstance", { enumerable: true, get: function () { return core_1.getInstance; } });
Object.defineProperty(exports, "listInstances", { enumerable: true, get: function () { return core_1.listInstances; } });
