// @osaas/client-core/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.waitForJobToComplete = exports.listJobs = exports.getJob = exports.removeJob = exports.createJob = exports.removeSubscription = exports.listSubscriptions = exports.waitForInstanceReady = exports.isValidInstanceName = exports.valueOrSecret = exports.instanceValue = exports.getInstanceHealth = exports.getLogsForInstance = exports.getPortsForInstance = exports.listInstances = exports.getInstance = exports.removeInstance = exports.createInstance = exports.Platform = exports.Context = exports.FetchError = exports.createFetch = exports.Log = void 0;
/** @module @osaas/client-core */
var log_1 = require("./log");
Object.defineProperty(exports, "Log", { enumerable: true, get: function () { return log_1.Log; } });
var fetch_1 = require("./fetch");
Object.defineProperty(exports, "createFetch", { enumerable: true, get: function () { return fetch_1.createFetch; } });
Object.defineProperty(exports, "FetchError", { enumerable: true, get: function () { return fetch_1.FetchError; } });
var context_1 = require("./context");
Object.defineProperty(exports, "Context", { enumerable: true, get: function () { return context_1.Context; } });
var platform_1 = require("./platform");
Object.defineProperty(exports, "Platform", { enumerable: true, get: function () { return platform_1.Platform; } });
var core_1 = require("./core");
Object.defineProperty(exports, "createInstance", { enumerable: true, get: function () { return core_1.createInstance; } });
Object.defineProperty(exports, "removeInstance", { enumerable: true, get: function () { return core_1.removeInstance; } });
Object.defineProperty(exports, "getInstance", { enumerable: true, get: function () { return core_1.getInstance; } });
Object.defineProperty(exports, "listInstances", { enumerable: true, get: function () { return core_1.listInstances; } });
Object.defineProperty(exports, "getPortsForInstance", { enumerable: true, get: function () { return core_1.getPortsForInstance; } });
Object.defineProperty(exports, "getLogsForInstance", { enumerable: true, get: function () { return core_1.getLogsForInstance; } });
Object.defineProperty(exports, "getInstanceHealth", { enumerable: true, get: function () { return core_1.getInstanceHealth; } });
Object.defineProperty(exports, "instanceValue", { enumerable: true, get: function () { return core_1.instanceValue; } });
Object.defineProperty(exports, "valueOrSecret", { enumerable: true, get: function () { return core_1.valueOrSecret; } });
Object.defineProperty(exports, "isValidInstanceName", { enumerable: true, get: function () { return core_1.isValidInstanceName; } });
Object.defineProperty(exports, "waitForInstanceReady", { enumerable: true, get: function () { return core_1.waitForInstanceReady; } });
var admin_1 = require("./admin");
Object.defineProperty(exports, "listSubscriptions", { enumerable: true, get: function () { return admin_1.listSubscriptions; } });
Object.defineProperty(exports, "removeSubscription", { enumerable: true, get: function () { return admin_1.removeSubscription; } });
var job_1 = require("./job");
Object.defineProperty(exports, "createJob", { enumerable: true, get: function () { return job_1.createJob; } });
Object.defineProperty(exports, "removeJob", { enumerable: true, get: function () { return job_1.removeJob; } });
Object.defineProperty(exports, "getJob", { enumerable: true, get: function () { return job_1.getJob; } });
Object.defineProperty(exports, "listJobs", { enumerable: true, get: function () { return job_1.listJobs; } });
Object.defineProperty(exports, "waitForJobToComplete", { enumerable: true, get: function () { return job_1.waitForJobToComplete; } });
__exportStar(require("./platform/maker"), exports);
__exportStar(require("./errors"), exports);
